### PR TITLE
Andy/database revisions

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**

### CODE REVIEW

File: go/libraries/doltcore/sqle/database_provider.go [hunk 1]
```diff
@@ -61,9 +115,201 @@ func (p DoltDatabaseProvider) AllDatabases() (all []sql.Database) {
...
+		Db:         db,
+		HeadCommit: cm,
+		WorkingSet: ws,
+		DbData: env.DbData{
+			Ddb: srcDb.ddb,
+			Rsw: static,
+			Rsr: static,
+			Drw: static,
+		},
+	}
```
Comment (Hydrocharged) [inline go/libraries/doltcore/sqle/database_provider.go:266]: This is such a mess lol. We should combine the three interfaces into another interface and reference that one instead. 6 separate copies of `static` is just ridiculous.

File: go/libraries/doltcore/sqle/dfunctions/dolt_commit.go [hunk 2]
```diff
@@ -60,8 +60,7 @@ func (d DoltCommitFunc) Eval(ctx *sql.Context, row sql.Row) (interface{}, error)
 	}
 
 	dSess := dsess.DSessFromSess(ctx.Session)
-
-	roots, ok := dSess.GetRoots(dbName)
+	roots, ok := dSess.GetRoots(nil, dbName)
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/sqle/dfunctions/dolt_commit.go:63]: Yeah this was just missed, good catch

File: go/libraries/doltcore/sqle/database_provider.go [hunk 3]
```diff
@@ -15,32 +15,83 @@
...
 	"github.com/dolthub/go-mysql-server/sql"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb"
+	"github.com/dolthub/dolt/go/libraries/doltcore/env"
+	"github.com/dolthub/dolt/go/libraries/doltcore/ref"
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle/dsess"
+)
+
+const (
+	dbRevisionDelimiter = "/"
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/sqle/database_provider.go:32]: always a forward slash, like in Git paths

File: go/libraries/doltcore/sqle/database_provider.go [hunk 4]
```diff
@@ -61,9 +115,201 @@ func (p DoltDatabaseProvider) AllDatabases() (all []sql.Database) {
...
+			return nil, false, err
+		}
+		return db, true, nil
+	}
+
+	return nil, false, nil
+}
+
+func (p DoltDatabaseProvider) RevisionDbState(ctx context.Context, revDB string) (dsess.InitialDbState, error) {
+	err := sql.ErrDatabaseNotFound.New(revDB)
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/sqle/database_provider.go:167]: good call

File: go/libraries/doltcore/sqle/database_provider.go (hunk 1 above)
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/sqle/database_provider.go:266]: This is more of a consequence of the mess that is the `sqle` data model. fixing that will fix this

File: go/libraries/doltcore/sqle/database_provider.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Hydrocharged) [reply go/libraries/doltcore/sqle/database_provider.go:32]: Ahh okay. Just wanted to make sure! Been burned on pathing stuff between operating systems so much...

**Review by Hydrocharged — APPROVED** [review]
` go/libraries/doltcore/sqle/dsess/session.go::LookupDbState()` still returns a redundant bool since it's always true if there's no error, but otherwise LGTM!

**Pull request merged**